optax/_src/base_test.py

```diff
@@ -59,5 +63,44 @@ def test_set_to_zero_is_stateless(self):
         self.variant(base.set_to_zero().init)(params=None), base.EmptyState())
 
 
+class StatelessTest(chex.TestCase):
+  """Tests for the stateless transformation."""
+
+  @chex.all_variants
+  @parameterized.parameters(False, True)
+  def test_stateless(self, on_leaves):
+    params = {'a': jnp.zeros((1, 2)), 'b': jnp.ones((1,))}
+    updates = {'a': jnp.ones((1, 2)), 'b': jnp.full((1,), 2.0)}
+
+    if on_leaves:
+      def weight_decay(g, p):
+        return g + 0.1 * p
+    else:
+      def weight_decay(g, p):
+        return jax.tree_map(lambda g_, p_: g_ + 0.1 * p_, g, p)
+
+    opt = base.stateless(weight_decay, on_leaves=on_leaves)
+    state = opt.init(params)
+    update_fn = self.variant(opt.update)
+    new_updates, new_state = update_fn(updates, state, params)
+    expected_updates = {'a': jnp.ones((1, 2)), 'b': jnp.array([2.1])}
+    chex.assert_trees_all_close(new_updates, expected_updates)
+
+      # Check repeated application
```

Comment:
Nit: Do you think it is necessary to check a repeated application? Since the transformation is stateless I think one application should suffice. While in theory testing more never hurts, it also makes this test longer (harder to see exactly which behaviour is being tested), less of a unit test that checks one behaviour, and might suggest that there is something more complicated going on in a repeated step.

Reply:
If you end up removing this, I think there would be so little boilerplate in the test that I'd also split it into two separate versions for with/without the tree map to keep things simple, especially if you have different function names rather than a boolean `on_leaves` flag.

Reply:
Don't think it's necessary; I removed the extra check.